src/lib.rs

```diff
@@ -232,6 +252,30 @@ impl<T> Tree<T> {
         self.vec.extend(other_tree.vec);
         unsafe { self.get_unchecked_mut(other_tree_root_id) }
     }
+
+    /// Maps a `Tree<T>` to `Tree<U>` by applying a function to all node values,
+    /// copying over the tree's structure and node ids untouched, consuming `self`.
+    pub fn map_values<U>(self, transform: impl Fn(T) -> U) -> Tree<U> {
```

Comment:
That would be my preferred API as well, but that would have required a couple of additional changes, which I was hesitant to make. From the top of my head:

1. We would need to expose a `IntoNodes<T>` iterator type, of course.
2. `Node<T>` would have to then also be made `pub` (since `IntoNodes<T>::Item = Node<T>`)

`Tree<T>`'s existing `IntoIterator` has `Item = T`, so a circuit via `tree.into_iter().collect()` would not work, since `FromIterator` (i.e. `.collect()`) would require `Item = Node<T>` instead. Which would be a bit surprising.

Also, once made `pub` it would feel sorta wrong for `Node<T>` to not also expose a couple of useful getter (and maybe even setters). Likewise as a user I'd now also expect to be able to get the value out of a `Node<T>`, such as via `node.into_value()`., etc.

That said I agree that providing mapping functionality in terms of iterator composition would be preferable. I just wasn't sure how far to go with this (and also hope for a timely follow-up release).

Reply:
The id/structure-preserving nature of `map_values()` is also important to me as it allows for then comparing mapped trees thanks to their shared node ids (at least as long as you don't mutate either of them afterwards).

Reply:
Another nail in the coffin of a safe `IntoIterator -> FromIterator` API for `Tree<T>`:

The "always contains at least a root node" invariant of `Tree<T>` is incompatible with the infallible signature of `FromIterator<Node<T>>`.

Reply:
> So maybe your two map variants as the way to go because they will always yield well-formed trees? Could we at least generalize them to filter_map to create holes?

I was considering this, too. But I wasn't sure how this would interact with the ids/indexes and my goal of "id/structure preserving". But maybe `filter_map()` simply doesn't preserve it, while `map()` does?

Reply:
Yeah, let's stay with your proven design. We can always add `filter_map` later on.

Reply:
Done.